tailfs/compositefs/compositefs.go

```diff
@@ -0,0 +1,261 @@
+	StatChildren bool
+	// Clock, if specified, determines the current time. If not specified, we
+	// default to time.Now().
+	Clock tstime.Clock
+}
+
+// New constructs a CompositeFileSystem that logs using the given logf.
+func New(opts Options) CompositeFileSystem {
+	logf := opts.Logf
+	if logf == nil {
+		logf = log.Printf
+	}
+	fs := &compositeFileSystem{
+		logf:         logf,
+		statChildren: opts.StatChildren,
+	}
+	if opts.Clock != nil {
+		fs.now = opts.Clock.Now
+	} else {
+		fs.now = time.Now
+	}
+	return fs
+}
+
+type compositeFileSystem struct {
+	logf         logger.Logf
+	statChildren bool
+	now          func() time.Time
+
+	// childrenMu guards children
+	childrenMu sync.Mutex
+	children   childrenByName
+}
+
+func (cfs *compositeFileSystem) AddChild(child *Child) {
+	cfs.childrenMu.Lock()
+	var oldChild *Child
+	existingChildPosition, existingChildFound := cfs.searchForChild(child.Name)
+	if existingChildFound {
+		oldChild = cfs.children[existingChildPosition]
+		cfs.children[existingChildPosition] = child
+	} else {
+		cfs.children = append(cfs.children, child)
+		sort.Sort(cfs.children)
+	}
+	cfs.childrenMu.Unlock()
+
+	if oldChild != nil {
+		closer, ok := oldChild.FS.(io.Closer)
+		if ok {
+			err := closer.Close()
+			if err != nil {
+				cfs.logf("failed to close child filesystem %v: %v", child.Name, err)
+			}
+		}
+	}
+}
+
+func (cfs *compositeFileSystem) RemoveChild(name string) {
+	cfs.childrenMu.Lock()
+	var oldChild *Child
+	existingChildPosition, existingChildFound := cfs.searchForChild(name)
+	if existingChildFound {
+		oldChild = cfs.children[existingChildPosition]
+		// shift children
+		for i := 0; i < len(cfs.children)-1; i++ {
+			if i > existingChildPosition {
+				cfs.children[i] = cfs.children[i+1]
+			}
+		}
+		// truncate children
+		cfs.children = cfs.children[:len(cfs.children)-1]
+	}
+	cfs.childrenMu.Unlock()
+
+	if oldChild != nil {
+		closer, ok := oldChild.FS.(io.Closer)
+		if ok {
+			err := closer.Close()
+			if err != nil {
+				cfs.logf("failed to close child filesystem %v: %v", name, err)
+			}
+		}
+	}
+}
+
+func (cfs *compositeFileSystem) SetChildren(children ...*Child) {
+	cfs.childrenMu.Lock()
+	oldChildren := cfs.children
+	cfs.children = childrenByName(children)
+	sort.Sort(cfs.children)
+	cfs.childrenMu.Unlock()
+
+	for _, child := range oldChildren {
+		closer, ok := child.FS.(io.Closer)
+		if ok {
+			_ = closer.Close()
+		}
+	}
+}
+
+func (cfs *compositeFileSystem) GetChild(name string) (webdav.FileSystem, bool) {
+	child, found := cfs.getChild(name)
+	if !found {
+		return nil, false
+	}
+	return child.FS, true
+}
+
+func (cfs *compositeFileSystem) getChild(name string) (*Child, bool) {
+	cfs.childrenMu.Lock()
+	defer cfs.childrenMu.Unlock()
+
+	childIndex, childExists := cfs.searchForChild(name)
+	if !childExists {
+		return nil, false
+	}
+	return cfs.children[childIndex], true
+}
+
+func (cfs *compositeFileSystem) searchForChild(name string) (int, bool) {
```

Comment:
Could we combine this and `getChild` with a single helper that does both? 

```go
func (cfs *CompositeFileSystem) findChildLocked(name string) (int, *Child, bool) {
   pos, ok := slices.BinarySearchFunc(cfs.children, name, func(c *Child, name string) int {
      return strings.Compare(c.Name, name)
   })
   if ok {
      return pos, cfs.children[pos], true
   }
   return pos, nil, false
}
```

Then we only have one ply of helpers rather than two, and only the exported methods need to deal with locking explicitly.

Reply:
Even if we don't combine them, I think it would be a good idea to be consistent about where we do locking. E.g., "exported methods acquire locks, helper methods assume locks are held" might be a good rubric.

Reply:
I like the `findChildLocked` simplification and will implement something along those lines. I don't like making the exported functions manage the mutex because a bunch of them use `pathInfoFor()` (formerly `getPathInfo()`), which needs to hold the mutex. The functions that call it don't otherwise need the mutex, so it would be more verbose and error prone to have them do it.